test/unit/range/view/view_async_buffer_test.cpp

```diff
@@ -0,0 +1,104 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <seqan3/alphabet/nucleotide/all.hpp>
+#include <seqan3/range/concept.hpp>
+#include <seqan3/range/view/async_buffer.hpp>
+#include <seqan3/range/view/single_pass_input.hpp>
+#include <seqan3/std/ranges>
+
+#include "../iterator_test_template.hpp"
+
+using namespace seqan3;
+
+using iterator_type = std::ranges::iterator_t<
+    decltype(std::declval<std::vector<seqan3::dna4>&>() | view::async_buffer(3))>;
+
+template <>
+struct iterator_fixture<iterator_type> : public ::testing::Test
+{
+    using iterator_tag = std::input_iterator_tag;
+    static constexpr bool const_iterable = false;
+
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+    std::vector<seqan3::dna4> expected_range{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+    decltype(view::async_buffer(vec, 3)) test_range = view::async_buffer(vec, 3);
+};
+
+using test_type = ::testing::Types<iterator_type>;
+INSTANTIATE_TYPED_TEST_CASE_P(iterator_fixture, iterator_fixture, test_type);
+
+
+TEST(async_buffer, in_out)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    auto v = vec | view::async_buffer(3);
+
+    EXPECT_TRUE(std::ranges::equal(vec, v));
+}
+
+TEST(async_buffer, buffer_size_zero)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    EXPECT_THROW(vec | view::async_buffer(0), std::invalid_argument);
+}
+
+TEST(async_buffer, buffer_size_huge)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    auto v = vec | view::async_buffer(100000);
+
+    EXPECT_TRUE(std::ranges::equal(vec, v));
+}
+
+TEST(async_buffer, destruct_with_full_buffer)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    auto v0 = vec | view::single_pass_input;
+
+    {
+        auto v1 = v0 | view::async_buffer(5);
+
+        // consume five elements (construction already consumes one)
+        auto b = std::ranges::begin(v1);
+        ++b; ++b; ++b; ++b;
+
+        // give time to rebuffer next five elements, not a problem if not successful
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+    } // thread sync at destruction of v1; tests working destruction with full buffer
+
+    EXPECT_GE(std::ranges::distance(v0), 17); // total of at most 10 chars consumed
```

Comment:
So this means we could basically remove the sleep command right

Reply:
I want to check that construction does not dead-lock when the buffer is not empty. The current code has a very high chance of testing this important behaviour. Removing the sleep means we have a very high change of *not* testing this behaviour.